src/is/L42/common/G.java

```diff
@@ -61,66 +97,87 @@ public G(Map<X, T> xInT, Set<X> vars) {
       Mdf.MutableFwd,Mdf.ImmutablePFwd)     
       );
     }
-  public boolean isVar(X x){return vars.contains(x);}
+  public boolean isVar(X x){
+    this.ensureFresh();
+    return vars.contains(x);
+    }
   public G keepVars(Collection<X> vars){
+    this.ensureFresh();
     Set<X> vs=new HashSet<>(this.vars);
     vs.retainAll(vars);
-    vs=Collections.unmodifiableSet(vs);
-    return new G(xInT,vs);
+    return new G(new HashMap<>(xInT),vs);
     }
   public T of(X x) {
+    this.ensureFresh();
     T res= xInT.get(x);
     assert res!=null:
       x;
     return res;
     }
   public T _of(X x) {
+    this.ensureFresh();
     T res= xInT.get(x);
     return res;
     }
   public T _of(Core.XP xp){
+    this.ensureFresh();
     if(xp instanceof Core.EX){return _of(((Core.EX)xp).x());}
     return ((Core.PCastT)xp).t();
     }
   public T of(Core.XP xp){
+    this.ensureFresh();
     if(xp instanceof Core.EX){return of(((Core.EX)xp).x());}
     return ((Core.PCastT)xp).t();
     }
+  private static record ToRemove(Set<X> keys, Set<X> vars) implements Consumer<G> {
+    @Override public void accept(G t) {
+      keys.forEach(k->t.xInT.remove(k));
+      vars.forEach(v->t.vars.remove(v));
+      } 
+    }
   public G plusEq(List<D> ds) {
+    this.ensureFresh();
     if(ds.isEmpty()){return this;}
-    Map<X,T> xInT=new HashMap<>(this.xInT);
-    Set<X> vars=new HashSet<>(this.vars);
+    Set<X> keysTR = new HashSet<>();
+    Set<X> varsTR = new HashSet<>();
     for(D di:ds){
       assert !xInT.containsKey(di.x()):di.x()+"; "+xInT;
       assert !vars.contains(di.x());
+      keysTR.add(di.x());
+      if(di.isVar()) varsTR.add(di.x());
       xInT.put(di.x(),di.t());
       if(di.isVar()){vars.add(di.x());}
       }
-    return new G(Collections.unmodifiableMap(xInT),Collections.unmodifiableSet(vars)); 
+    return this.newChild(new ToRemove(keysTR, varsTR));
     }
   public G plusEq(X x, T t) {
-    Map<X,T> xInT=new HashMap<>(this.xInT);
+    this.ensureFresh();
     assert !xInT.containsKey(x):x+" in "+xInT;
     xInT.put(x,t);
-    return new G(Collections.unmodifiableMap(xInT),vars); 
+    return this.newChild(new ToRemove(Set.of(x), Set.of()));

```

Comment:
This may be faster and logically better by using just a consumer doing the removing x action.
Otherwise, why we use the consumer abstraction?